backend/controller/identity/identity.go

```diff
@@ -0,0 +1,182 @@
+	err := s.verifier.Verify(signedData)
+	if err != nil {
+		return fmt.Errorf("failed to verify data: %w", err)
+	}
+
+	return nil
+}
+
+func (s Service) getKeyPair(ctx context.Context) (internalidentity.KeyPair, error) {
+	identity, err := s.dal.GetOnlyIdentityKey(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get only identity key: %w", err)
+	}
+
+	reader := keyset.NewBinaryReader(bytes.NewReader(identity.Private.Bytes()))
+	aead, err := s.encryption.AEAD()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get AEAD: %w", err)
+	}
+
+	handle, err := keyset.Read(reader, aead)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read keyset: %w", err)
+	}
+
+	keyPair := internalidentity.NewTinkKeyPair(*handle)
+	return keyPair, nil
+}
+
+const verificationText = "My voice is my passport, verify me."
+
+func (s Service) ensureIdentity(ctx context.Context) (err error) {
+	logger := log.FromContext(ctx)
+	tx, err := s.dal.Begin(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to begin transaction: %w", err)
+	}
+	defer tx.CommitOrRollback(ctx, &err)
+
+	_, err = s.dal.GetOnlyIdentityKey(ctx)
+	if err != nil {
+		if !libdal.IsNotFound(err) {
+			return fmt.Errorf("failed to get only identity key: %w", err)
+		}
+
+		logger.Debugf("Generating identity key")
+		err = s.generateAndSaveIdentity(ctx, tx)
+		if err != nil {
+			return fmt.Errorf("failed to generate and save identity: %w", err)
+		}
+	} else {
+		logger.Debugf("Identity key already exists")
+	}
+
+	return nil
+}
+
+func (s Service) generateAndSaveIdentity(ctx context.Context, tx *dal.DAL) error {
+	pair, err := internalidentity.GenerateTinkKeyPair()
+	if err != nil {
+		return fmt.Errorf("failed to generate key pair: %w", err)
+	}
+
+	signer, err := pair.Signer()
+	if err != nil {
+		return fmt.Errorf("failed to create signer: %w", err)
+	}
+
+	signed, err := signer.Sign([]byte(verificationText))
+	if err != nil {
+		return fmt.Errorf("failed to sign verification: %w", err)
+	}
+
+	verifier, err := pair.Verifier()
+	if err != nil {
+		return fmt.Errorf("failed to create verifier: %w", err)
+	}
+
+	// For total sanity, verify immediately
+	if err = verifier.Verify(*signed); err != nil {
+		return fmt.Errorf("failed to verify signed verification: %w", err)
+	}
+
+	// TODO: Make this support different encryptors.
+	// Might need to refactor internal/identity to access controller encryption types.
+	// It's a bit tricky because you can't take out the private key from the keyset without
+	// encrypting it with the AEAD.
+	handle := pair.Handle()
+	buf := new(bytes.Buffer)
+	writer := keyset.NewBinaryWriter(buf)
+	aead, err := s.encryption.AEAD()
+	if err != nil {
+		return fmt.Errorf("failed to get AEAD: %w", err)
+	}
+	if err := handle.Write(writer, aead); err != nil {
+		return fmt.Errorf("failed to write keyset: %w", err)
+	}
+	encryptedIdentityColumn := api.EncryptedIdentityColumn{}
+	encryptedIdentityColumn.Set(buf.Bytes())
```

Comment:
Is this encrypted or plain text? If it's plain text it shouldn't be in an `EncryptedIdentityColumn`?

Reply:
buf.Bytes() is actually encrypted already!

Reply:
Tink strongly prefers (and thus makes it the default) to export / write keysets in encrypted form only. The Write operation on line 162 takes the AEAD primitive and uses it to encrypt the keyset. This encrypted keyset is FLE encrypted again via our EncryptedColumen abtraction